### ISSUE TITLE
Supervisor restart should allow for exponential backoff on retry

### ISSUE THREAD

**Issue by rtyler** [bastion-rs/bastion#146]
**Is your feature request related to a problem? Please describe.**

For some failure models, such as a client trying to connect to a server which is offline, the current supervision model will effectively `loop { connect(); }` over and over again, with no "safe" backoff on retrying the child.

**Describe the solution you'd like**

Ideally, I would like to specify an ExponentialBackOff strategy for my supervisor, such that when children hit failures, the process will not simply consume 100% CPU trying to retry the child.

[backoff](https://crates.io/crates/backoff) is an example implementation that works on arbitrary closures, which may be worth borrowing code or ideas from.

**Describe alternatives you've considered**

Me manually putting a thread sleep into the child doesn't seem like a good idea :smile_cat:

**Comment by rtyler** [comment]
My temporary workaround is to add a `map_err` to my workers where I invoke `connect` (a predictable source of errors):


```rust
                let (mut socket, response) =
                    connect(Url::parse("ws://localhost:8000/ws/").unwrap())
                    .map_err(|e| {
                        error!("Failed to connect!");
                        sleep(Duration::from_secs(1));
                    }).expect("Can't connect");
```

**Comment by vertexclique** [comment]
This is now implemented in #157 . Closing… :)
